### PR TITLE
ARROW-5092: [C#] Create a dummy .git directory to download the source files from GitHub with Source Link

### PR DESCRIPTION
**Pull request by shiro615**

### CODE REVIEW

File: dev/release/post-06-csharp.sh
Comment (eerhardt) [inline dev/release/post-06-csharp.sh]: I think this `.git` directory needs to be one directory up, above the `csharp` directory. Basically, it needs to match what a "normal" git repo would look like.

File: dev/release/post-06-csharp.sh
Comment (eerhardt) [inline dev/release/post-06-csharp.sh]: When `dotnet pack` is executed, it needs to be executed from the `${archive_name}/csharp` directory. But it looks like the script is currently in the `${archive_name}/csharp/.git` directory. Make sure you change to the correct directory before running `dotnet pack`.

**Review by eerhardt — COMMENTED** [review]
Thanks @shiro615 for doing this work. It will really help out the .NET community that uses this package.

There is one more change that is needed in order to fix the issue. There was an update on the .NET tooling side (See https://github.com/dotnet/sourcelink/pull/266). In order to pick up that change, can you update this version number as well?

https://github.com/apache/arrow/blob/dfb9e7af3cd92722893a3819b6676dfdef08f896/csharp/src/Apache.Arrow/Apache.Arrow.csproj#L17

Update that to version `1.0.0-beta2-19270-01`. I've tested it out locally, and with your script, and my recommended changes, this will fix the problem.

**Review by kou — CHANGES_REQUESTED** [review]
Can we do this in `dev/release/02-source.sh`? Then we can verify this in release vote process.

**Comment by kou** [comment]
Dummy `.git/` in release archive may break Python package because it check `.git/` directory...
https://github.com/apache/arrow/blob/master/python/setup.py#L512

**Comment by shiro615** [comment]
@eerhardt Thank you for pointing out. I'll fix them.

**Comment by shiro615** [comment]
> Can we do this in dev/release/02-source.sh? Then we can verify this in release vote process.

Yes. I've moved  the script to create a dummy `.git` directory to `dev/release/02-source.sh`.

> Dummy .git/ in release archive may break Python package because it check .git/ directory...

How about checking files that do not exist in the dummy `.git` directory such as `.git/index`?

```
--- a/python/setup.py
+++ b/python/setup.py
@@ -509,7 +509,7 @@ def _move_shared_libs_unix(build_prefix, build_lib, lib_name):
 # If the event of not running from a git clone (e.g. from a git archive
 # or a Python sdist), see if we can set the version number ourselves
 default_version = '0.14.0-SNAPSHOT'
-if (not os.path.exists('../.git')
+if (not os.path.exists('../.git/index')
         and not os.environ.get('SETUPTOOLS_SCM_PRETEND_VERSION')):
```

File: dev/release/02-source.sh
Comment (kou) [inline dev/release/02-source.sh]: Can we `release_hash` here?

Comment (shiro615) [reply dev/release/02-source.sh]: I've used `${release_hash}`.

**Comment by shiro615** [comment]
@eerhardt 

Is there any way to check if the Source Link information is set correctly in the symbol file?
For example, the following information.

```
1: '/home/kou/work/cpp/arrow.kou/apache-arrow-0.13.0/csharp/src/Apache.Arrow/Flatbuf/FlatBuffers/ByteBuffer.cs' (#19c)            C# (#3)   SHA-1 (#2)     04-64-A0-48-82-EA-F5-B5-50-EC-CA-9F-85-75-E2-95-A4-EC-AB-B3 (#1b7)   
2: '/home/kou/work/cpp/arrow.kou/apache-arrow-0.13.0/csharp/src/Apache.Arrow/Flatbuf/FlatBuffers/ByteBufferUtil.cs' (#68f)        C# (#3)   SHA-1 (#2)     F0-4F-28-53-88-A4-E0-6E-F1-1F-17-F6-CD-FE-0E-64-AB-0B-C2-95 (#6aa)   
```

I'm trying to add a test that the following information is set correctly in `csharp/artifacts/Apache.Arrow/Release/Apache.Arrow.nuspec`.

 ```
<repository type="git" url="https://github.com/apache/arrow" commit="..." 
```
If we can check the Source Link information I want to add the test too.

**Comment by eerhardt** [comment]
> Is there any way to check if the Source Link information is set correctly in the symbol file?

Yes you can with the following steps:

```
dotnet tool install -g sourcelink

sourcelink test $(RepoRoot)/csharp/artifacts/Apache.Arrow/Release/netstandard1.3/Apache.Arrow.pdb
sourcelink test $(RepoRoot)/csharp/artifacts/Apache.Arrow/Release/netcoreapp2.1/Apache.Arrow.pdb
```

See https://github.com/ctaggart/SourceLink#test

**Comment by shiro615** [comment]
Thanks! It worked. I'll add the test for `dev/02-source-test.rb`.

File: dev/release/02-source-test.rb
Comment (shiro615) [inline dev/release/02-source-test.rb]: @kou I divided this file into a separate file to test only this script.
How do you think about this?

**Review by eerhardt — APPROVED** [review]
Looks good from the C# side. I'll let the maintainers comment on the `dev/release` changes.

File: dev/release/02-source-test.rb
Comment (kou) [reply dev/release/02-source-test.rb]: Selecting target logics by environment variable is better than splitting a script.
See also:

  * https://github.com/apache/arrow/blob/master/dev/release/00-prepare.sh#L159-L165
  * https://github.com/apache/arrow/blob/master/dev/release/verify-release-candidate.sh#L423-L434

Comment (shiro615) [reply dev/release/02-source-test.rb]: Thanks! I will run `dev/release/02-source.sh` in test cases.

**Comment by shiro615** [comment]
@kou  Can you review this?
CI failure is unrelated.

File: dev/release/02-source.sh
Comment (kou) [inline dev/release/02-source.sh]: Could you put all codes after `set -e` to check all codes by `set -e`?

File: dev/release/02-source.sh
Comment (kou) [inline dev/release/02-source.sh]: Is `_TAG` needed? It seems that we don't prepare tag for C GLib.

File: dev/release/02-source.sh
Comment (kou) [inline dev/release/02-source.sh]: We can already override `release_hash` by `release_hash=XXX dev/release/02-prepare.sh`. So we don't need this.

File: dev/release/02-source.sh
Comment (kou) [reply dev/release/02-source.sh]: `SOURCE_` prefix is better.

File: dev/release/02-source.sh
Comment (kou) [inline dev/release/02-source.sh]: How about `SOURCE_UPLOAD`?

File: dev/release/02-source.sh
Comment (kou) [inline dev/release/02-source.sh]: How about move RAT check outside of `PREPARE_RELEASE_TAG`?
RAT check is useful.
How about adding `SOURCE_RAT`?

File: dev/release/02-source-test.rb
Comment (kou) [inline dev/release/02-source-test.rb]: Could you use `assert do end` for predicate?

```ruby
assert do
  File.exist?("dist/pyarrow-#{@release_version}a0.tar.gz")
end
```

File: dev/release/02-source-test.rb
Comment (kou) [inline dev/release/02-source-test.rb]: We should use `sh(COMMAND, ARG1, ARG2, ...)` style instead of `sh(COMMAND_LINE)` to avoid shell injection.
Exception: We need to use `COMMAND_LINE` style when we need shell feature such as pipe and redirect.

File: dev/release/02-source-test.rb
Comment (kou) [inline dev/release/02-source-test.rb]: `File.read("Apache.ARrow.nuspec")` is better because this close file descriptor immediately.
`open().read` close file descriptor on GC.

File: dev/release/02-source-test.rb
Comment (kou) [inline dev/release/02-source-test.rb]: Could you put a space around "=>"? `{"PREPARE_DEFAULT" => "0"}`

**Review by kou — CHANGES_REQUESTED** [review]
Thanks!
I'll try this locally later.

**Comment by kou** [comment]
The CI failure has been fixed by #4650.

File: dev/release/02-source.sh
Comment (shiro615) [reply dev/release/02-source.sh]: I've moved all codes after `set -e`.

File: dev/release/02-source.sh
Comment (shiro615) [reply dev/release/02-source.sh]: The name of this environment variables was not appropriate. I've removed `_TAG`.

File: dev/release/02-source.sh
Comment (shiro615) [reply dev/release/02-source.sh]: I've set `release_hash` in `02-source-test.rb`.

File: dev/release/02-source.sh
Comment (shiro615) [reply dev/release/02-source.sh]: I've added `SOURCE_RAT`.

File: dev/release/02-source-test.rb
Comment (shiro615) [reply dev/release/02-source-test.rb]: I've fixed it.

File: dev/release/02-source-test.rb
Comment (shiro615) [reply dev/release/02-source-test.rb]: I see.  I've used `sh(COMMAND, ARG1, ARG2, ...) style`.
```
home = ENV['HOME']
assert do
  sh("#{home}/.dotnet/tools/sourcelink",
  "test",
  "artifacts/Apache.Arrow/Release/netcoreapp2.1/Apache.Arrow.pdb")
end
```

File: dev/release/02-source-test.rb
Comment (shiro615) [reply dev/release/02-source-test.rb]: I've used `File.read()`.

File: dev/release/02-source-test.rb
Comment (shiro615) [reply dev/release/02-source-test.rb]: I've fixed it.

**Comment by shiro615** [comment]
By the way, macOS seems not to support [cp -r -L](https://github.com/apache/arrow/blob/master/dev/release/02-source.sh#L74).
`cp: the -H, -L, and -P options may not be specified with the -r option.`

macOS users need to install `coreutils` by `brew uninstall coreutils`. I'll add the description in https://cwiki.apache.org/confluence/display/ARROW/Release+Management+Guide.

**Review by kou — APPROVED** [review]
Thanks!

I've pushed some changes:

* `cp`: Use `cp -R` instead of `cp -r`. `cp -R` works on macOS.
* Remove `sourcelink test` test. Because it fails when we don't the current commit to the repository yet. It should be done in `verify-release-candidate.sh`. Do you want to work on this as another pull request?
* Add more tests.

We can merge this when CI is green.

**Comment by shiro615** [comment]
Thanks! 

> Remove sourcelink test test. Because it fails when we don't the current commit to the repository yet. It should be done in verify-release-candidate.sh. Do you want to work on this as another pull request?

Yes. I'll work on this. I've created https://issues.apache.org/jira/browse/ARROW-5695.

**Comment by shiro615** [comment]
CI is green. I'll merge this.

**Comment by kou** [comment]
I'm working on #4613. I found that this causes Java failure in release verification script.
I'll change `02-source.sh` to create `csharp/dummy.git` instead of `.git` and `post-06-csharp.sh` to move `csharp/dummy.git` to `.git` around creating C# package in #4613.

**Comment by shiro615** [comment]
Oh...sorry.  I didn't verify the script.

**Pull request closed without merging**
